pkg/master/thirdparty_controller.go

```diff
@@ -0,0 +1,109 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/expapi"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	thirdpartyresourceetcd "k8s.io/kubernetes/pkg/registry/thirdpartyresource/etcd"
+	"k8s.io/kubernetes/pkg/registry/thirdpartyresourcedata"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+const thirdpartyprefix = "/thirdparty/"
+
+func makeThirdPartyPath(group string) string {
+	return thirdpartyprefix + group
+}
+
+type APIInterface interface {
+	// Remove a third party resource based on the RESTful path for that resource
+	RemoveThirdPartyResource(path string) error
+	// Install a third party resource described by 'rsrc'
+	InstallThirdPartyResource(rsrc *expapi.ThirdPartyResource) error
+	// Is a particular third party resource currently installed?
+	HasThirdPartyResource(rsrc *expapi.ThirdPartyResource) (bool, error)
+	// List all currently installed third party resources
+	ListThirdPartyResources() []string
+}
+
+type ThirdPartyController struct {
+	master                     APIInterface
+	thirdPartyResourceRegistry *thirdpartyresourceetcd.REST
+}
+
+func (t *ThirdPartyController) SyncOneResource(rsrc *expapi.ThirdPartyResource) error {
+	hasResource, err := t.master.HasThirdPartyResource(rsrc)
+	if err != nil {
+		return err
+	}
+	if !hasResource {
+		return t.master.InstallThirdPartyResource(rsrc)
+	}
+	return nil
+}
+
+func (t *ThirdPartyController) SyncResources() error {
+	list, err := t.thirdPartyResourceRegistry.List(api.NewDefaultContext(), labels.Everything(), fields.Everything())
+	if err != nil {
+		return err
+	}
+	return t.syncResourceList(list)
+}
+
+func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
+	existing := sets.String{}
+	switch list := list.(type) {
+	case *expapi.ThirdPartyResourceList:
+		for ix := range list.Items {
+			item := &list.Items[ix]
+			_, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(item)
+			if err != nil {
+				return err
+			}
+			existing.Insert(makeThirdPartyPath(group))
+			if err := t.SyncOneResource(item); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("expected a *ThirdPartyResourceList, got %#v", list)
+	}
+	installed := t.master.ListThirdPartyResources()
+	for _, installedAPI := range installed {
+		found := false
+		for _, apiPath := range existing.List() {
```

Comment:
When we add APIs to the master we add sub-paths for them.  e.g. adding `/api/company.com/foo` also adds a path `/api/company.com/foo/v1`

So we have to test prefixes and we have no visibility into what actually was created.

We could add more tracking inside of the master to understand this, but it's actually as pretty significant refactor of the master code.